R/drive_list.R

```diff
@@ -0,0 +1,162 @@
+#' List files on Google Drive
+#'
+#' @param path character vector, path name(s) where the Google drive files are
+#'   that you would like to list. Defaults to the "My Drive" directory.
+#' @param pattern character vector, regular expression(s) of title(s) of
+#'   documents to output in a tibble. If it is `NULL` (default), information
+#'   about all documents in drive will be output in a tibble.
+#' @param ... name-value pairs to query the API
+#' @param verbose logical, indicating whether to print informative messages
+#'   (default `TRUE`)
+#'
+#'   This will default to the most recent 100 files on your Google Drive. If you
+#'   would like more than 100, include the `pageSize` parameter. For example, if
+#'   I wanted 200, I would run `drive_ls(pageSize = 200)`.
+
+#' Helpful links for forming queries:
+#'   * <https://developers.google.com/drive/v3/web/search-parameters>
+#'   * <https://developers.google.com/drive/v3/reference/files/list>
+
+#' @return tibble containing the name, type, and id of files on your google
+#'   drive (default 100 files)
+#' @examples
+#' \dontrun{
+#' ## list user's Google Sheets
+#' drive_list(q = "mimeType='application/vnd.google-apps.spreadsheet'")
+#' }
+#'
+#' @export
+drive_list <- function(path = NULL, pattern = NULL, ..., verbose = TRUE){
+
+  folder <- NULL
+  if (!is.null(path)){
+    folder_pattern <- paste0("^", gsub("/", "$|^", path), "$")
+    folder_nms <- unlist(strsplit(path, "/"))
+    folder_order <- tibble::tibble(name = folder_nms,
+                                   dir = 1:length(folder_nms))
+    folder_tbl <- drive_list(pattern = folder_pattern,
+                             fields = paste0("files/parents,files/name,files/mimeType,files/id"),
+                             q = "mimeType='application/vnd.google-apps.folder'")
+
+    if (!all(folder_nms %in% folder_tbl$name)){
+      spf("We could not find the file path '%s' on your Google Drive", path)
+    }
+
+    # merge in to get the correct directory order & sort, oh how I miss the %>% :(
+    folder_tbl <- tibble::as_data_frame(merge(folder_tbl, folder_order, by = "name"))
+    folder_tbl_sort <- folder_tbl[order(folder_tbl$dir), ]
+
+    parent_id <- root_folder()
+    keep_folders <- NULL
+    for (i in 1:nrow(folder_tbl_sort)){
+      subfolder <- folder_tbl_sort[i, ]
+      keep <- subfolder$folder_id == parent_id
+      if (keep){
+        parent_id <- subfolder$id
+      }
+      keep_folders[i] <- keep
+    }
+
+    folder_tbl_keep <- folder_tbl_sort[keep_folders, ]
+    leafmost <- nrow(folder_tbl_keep)
+    if (leafmost != length(folder_nms)){
+      spf("Uh oh, something went wrong. We could not find the path '%s' on your Google Drive", path)
```

Comment:
This could happen in the case that you have something like `path = "foo/bar/yo"` where all of those folders exist somewhere on your drive, but not in the nested way as suggested (for example the real path would be `foo/bar/baz/yo`). In this case, no errors would be thrown up until this point because all of the folders exist, but when we try to find folders within the folders, only `foo` and `bar` would come out of this function since `yo` is not in `bar`. I've [updated to be more informative](https://github.com/tidyverse/googledrive/blob/drive_list/R/drive_list.R#L125-L128).